resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java

```diff
@@ -289,4 +278,90 @@ public String toString() {
             }
         }
     }
+
+    private static final class Entries {
+        private final ReadWriteLock lock = new ReentrantReadWriteLock();
+        private final Lock readLock = lock.readLock();
+        private final Lock writeLock = lock.writeLock();
+
+        // Access to this field is guarded by the ReadWriteLock.
+        //
+        // It's important that this List will never be modified itself but just replaced when an new entry is added,
+        // or removed.
+        private List<DefaultDnsCacheEntry> entries;
+
+        Entries(DefaultDnsCacheEntry entry) {
+            entries = Collections.unmodifiableList(Collections.singletonList(entry));
+        }
+
+        List<DefaultDnsCacheEntry> get() {
+            readLock.lock();
+            try {
+                return entries;
+            } finally {
+                readLock.unlock();
+            }
+        }
+
+        void add(DefaultDnsCacheEntry e) {
+            if (e.cause() == null) {
+                writeLock.lock();
+                try {
+                    if (!entries.isEmpty()) {
+                        final DefaultDnsCacheEntry firstEntry = entries.get(0);
+                        if (firstEntry.cause() != null) {
+                            assert entries.size() == 1;
+                            firstEntry.cancelExpiration();
+                            entries = Collections.unmodifiableList(Collections.singletonList(e));
+                            return;
+                        }
+                    }
+
+                    List<DefaultDnsCacheEntry> newEntries = new ArrayList<DefaultDnsCacheEntry>(entries.size() + 1);
+                    newEntries.addAll(entries);
+                    newEntries.add(e);
+                    entries = Collections.unmodifiableList(newEntries);
+                } finally {
+                    writeLock.unlock();
+                }
+            } else {
+                List<DefaultDnsCacheEntry> oldEntries;
+
+                writeLock.lock();
+                try {
+                    oldEntries = entries;
```

Comment:
if this field is treated as copy on write ... can we just use a volatile and avoid the locking? this would make things much more clear because we let the list escape the lock above in `get()`.

Reply:
Yeah I think this would work...  Let me do this